crypto/ec/ecdsa_ossl.c

```diff
@@ -100,13 +100,16 @@ static int ecdsa_sign_setup(EC_KEY *eckey, BN_CTX *ctx_in,
         /*
          * We do not want timing information to leak the length of k, so we
          * compute G*k using an equivalent scalar of fixed bit-length.
+         *
+         * We unconditionally perform both of these additions to prevent a
+         * small timing information leakage.  We then choose the sum that is
+         * BN_num_bits(order) + 1 bits in length.  This guarantees the code
+         * path used in the constant time implementations elsewhere.
          */
-
-        if (!BN_add(k, k, order))
+        if (!BN_add(r, k, order)
+            || !BN_add(X, r, order)
+            || !BN_copy(k, BN_num_bits(r) > BN_num_bits(order) ? r : X))
             goto err;
-        if (BN_num_bits(k) <= BN_num_bits(order))
-            if (!BN_add(k, k, order))
-                goto err;
```

Comment:
Why doesn't setting most significant bit work here?

Reply:
Doing so broke one of the tests.
Setting the bit changes the value of k mod order, doing additions doesn't.

Reply:
I didn't want to modify the tests.

Reply:
I don't follow. `k` is just random number, and if you permit values larger than `order`, then setting most significant bit would be equivalent of choosing another random number, k | msb mod order? Do note that I'm talking about most significant bit, i.e. BN_num_bits(order) - 1...

Reply:
Concern is basically that the kind of conditional copy that is suggested is not constant time. Well, it's better than before, but question is if one does it, then why stop here...

Reply:
The problem was that one of the tests replaces the RNG so it outputs known "random" bytes.  Then does an expected value test afterwards.  This test failed.  Rather than changing the test, it seemed safer to keep the k value constant modulo order.

Reply:
I am for changing the test case and the approach.  This seems like it would be better done after the release.

Reply:
The test in question is `x9_62_tests` in `ecdsatest.c`.